[JPA] 2단계 - 연관 관계 매핑

jaewanLee opened the pull request:
안녕하세요 준일님! 앞으로 잘부탁드립니다.
1단계 리뷰해주신 내용 반영했습니다. 친절한 리뷰 감사드립니다!

2단계 연관관계 매핑 완료했습니다.
리뷰 요청드립니다.

jaewanLee committed:
feat : 피드백 반영 Entity 개행

jaewanLee committed:
feat : 피드백 반영 BaseEntity로 id 이동

jaewanLee committed:
feat : 피드백 반영 @Column 속성으로 unique key 생성
- User: userId

jaewanLee committed:
feat : 연관관계 매핑
- Answer: writer(User), question(Question)
- DeleteHistory: deletedBy(User)
- Question: writer(User)

jaewanLee committed:
fix : entity 수정으로인한 Repository test 수정

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -9,13 +9,18 @@
-    private Long writerId;
-    private Long questionId;
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_answer_writer"))
+    @OneToOne
+    private User writer;
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+    @ManyToOne
+    private Question question;
```
LazyLoad 설정도 같이 해주면 어떨까요?

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -9,13 +9,18 @@
+    private User writer;
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+    @ManyToOne
```
`@Where` 어노테이션을 이용해서 조건에 맞는 엔티티만 매핑할 수도 있답니다!

JunilHwang commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -1,16 +1,29 @@
+import javax.persistence.*;
 import java.util.Date;
 
 @MappedSuperclass
 public abstract class BaseEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
     @Column(nullable = false)
     @Temporal(TemporalType.TIMESTAMP)
     private Date createdAt = new Date();
+
     @Temporal(TemporalType.TIMESTAMP)
     private Date updatedAt;
```
`@EntityListeners(AuditingEntityListener.class)` 어노테이션을 활용해보세요!
더불어서 `@CreatedDate`, `@LastModifiedDate` 도 같이 달아주면 좋을 것 같습니다.

참고링크: https://edu.nextstep.camp/s/ffeVDScX/ls/dsH0WNPf

JunilHwang commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -7,21 +7,28 @@
-    private Long deletedById;
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_delete_history_to_user"))
+    @OneToOne
+    private User deletedBy;
```
잘 생각해보면 `@OneToOne`이 아니라 `@ManyToOne`이 아닐까요?

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -9,13 +9,18 @@
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_answer_writer"))
+    @OneToOne
+    private User writer;
```
잘 생각해보면 `@OneToOne`이 아니라 `@ManyToOne`이 아닐까요?

JunilHwang commented on src/main/java/qna/domain/Question.java:
```diff
@@ -5,48 +5,45 @@
+
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_question_writer"))
+    @OneToOne
+    private User writer;
```
잘 생각해보면 `@OneToOne`이 아니라 `@ManyToOne`이 아닐까요?

JunilHwang reviewed (changes requested):
안녕하세요 재완님! 연관 관계 매핑 미션 잘 수행해주셨네요 👍 
리뷰를 요약하자면 다음과 같습니다.

- LazyLoad 설정 저용해보기
  - 참고링크: https://ict-nroo.tistory.com/132
- OneToOne -> ManyToOne으로 변경해보기 
  -  OneToOne은 말 그대로 1:1 관계입니다. 가령 글 작성자는 한 명이지만, 글은 여러개일 수 있겠죠? 1:1이라는건, 글 작성자도 한 명이고, 작성자의 글도 한 개라는 의미입니다.
- `@EntityListeners(AuditingEntityListener.class)` 활용해보기

위와 같습니다. 확인 후 다시 피드백 요청 부탁드려요 🙇‍♂️

jaewanLee committed:
feat: LazyLoading 설정 적용
- Answer: question
- DeleteHistory: deletedBy
- Question: writer

jaewanLee committed:
feat: EntityListener 적용
- @CreatedDate 적용
- @LastModifiedDate 적용

jaewanLee committed:
fix: 연관관계 매핑 수정
- Answer-writer: OneToOne -> ManyToOne
- DeleteHistory-deletedBy: OneToOne -> ManyToOne
- Question-writer: OneToOne -> ManyToOne

jaewanLee committed:
feature: @Where 어노테이션 사용

JunilHwang reviewed (approved):
안녕하세요 재완님! 피드백 반영 빠르게 잘 해주셨네요 👍 
Step2는 여기서 마무리하겠습니다.
바로 Step3 진행해주세요!
